pyspinw/calculations/spinwave.py

```diff
@@ -26,158 +28,312 @@ class Coupling:
 
+    return energies
 
-        try:
-            sqrt_hamiltonian = np.linalg.cholesky(hamiltonian_matrix)
-            method = CalculationMethod.CHOLESKY
 
-        except np.linalg.LinAlgError: # Catch postive definiteness errors
-            # l, d, perm = ldl(hamiltonian_matrix) # To LDL^\dagger (i.e. adjoint on right)
-            l, d, _ = ldl(hamiltonian_matrix) # To LDL^\dagger (i.e. adjoint on right)
-            sqrt_hamiltonian = l @ np.sqrt(d)
+def spinwave_calculation(
+    rotations: list[np.ndarray],
+    magnitudes: np.ndarray,
+    q_vectors: np.ndarray,
+    couplings: list[Coupling],
+    positions: list[np.ndarray],
+    field: MagneticField | None = None,
+    save_sab: bool = False,
+) -> (np.ndarray, np.ndarray, Optional[np.ndarray]):
+    """Calculate the energies and spin-spin correlation for a set of q-vectors."""
+    C, z, spin_coefficients, Az = _calc_q_independent(rotations, magnitudes, couplings, field)
+    n_sites = len(rotations)
+
+    # Linear algebra routines in numpy are already parallelised and usually use 4 cores
+    # for a single process, so we want to reduce contention by using fewer processes.
+    n_proc = max(int(np.floor(multiprocessing.cpu_count() / 4)), 1)
+    with ProcessPoolExecutor() as executor:
+        q_calculations = [
+            executor.submit(
+                _calc_chunk_spinwave, q, C, n_sites, z, spin_coefficients, couplings, positions, Az, save_sab
+            )
+            for q in _get_q_chunks(q_vectors, n_proc)
+        ]
+    wait(q_calculations)
+    results = [future.result() for future in q_calculations]
+    energies = np.concat(tuple(result[0] for result in results))
+    intensities = np.concat([result[1] for result in results])
+    if save_sab:
+        sab = [result[2] for result in results]
+        return energies, intensities, sab
+
+    return energies, intensities
+
+
+def _calc_chunk_spinwave(
+    q_vectors: np.ndarray,
+    C: np.ndarray,
+    n_sites: int,
+    z: np.ndarray,
+    spin_coefficients: np.ndarray,
+    couplings: list[Coupling],
+    positions: list[np.ndarray],
+    Az: np.ndarray | None = None,
+    save_sab: bool = False,
+) -> (np.ndarray, np.ndarray, Optional[np.ndarray]):
+    """Calculate the energies and S'^alpha,beta for a chunk of q-values."""
+    energies = []
+    intensities = []
+    sabs = []
 
-            # TODO: Check for actual diagonal (could potentially contain non-diagonal 2x2 blocks)
-            method = CalculationMethod.LDL
+    for q in q_vectors:
+        sqrt_hamiltonian = _calc_sqrt_hamiltonian(q, C, n_sites, z, spin_coefficients, couplings, Az)
 
         sqrt_hamiltonian_with_commutation = sqrt_hamiltonian.copy()
         sqrt_hamiltonian_with_commutation[n_sites:, :] *= -1  # This is C*K
 
         to_diagonalise = np.conj(sqrt_hamiltonian).T @ sqrt_hamiltonian_with_commutation
 
-        energies.append(np.linalg.eigvals(to_diagonalise))
+        eigvals, eigvecs = np.linalg.eig(to_diagonalise)

```

Comment:
In this method we always assume that the part to be diagonalised is Hermitian - and to make the `antiferro_chain` test pass we need to use `eigh` here otherwise the eigenvectors are not always orthogonalised and will disagree with the Rust computation.

In addition, in order to pass the `antiferro_chain` test (see comment on Rust code below), we need to add a small delta to the diagonal, changing the line to:

```suggestion
        eigvals, eigvecs = np.linalg.eigh(to_diagonalise + np.diag(np.array(range(to_diagonalise.shape[0]))*1e-12))
```